src/descriptor/tr.rs

```diff
@@ -0,0 +1,671 @@
+// Tapscript
+
+use crate::policy::semantic::Policy;
+use crate::policy::Liftable;
+use crate::util::{varint_len, witness_size, MsKeyBuilder};
+use crate::{DescriptorTrait, ForEach, ForEachKey, Satisfier, ToPublicKey, TranslatePk};
+
+use super::checksum::{desc_checksum, verify_checksum};
+use bitcoin::blockdata::opcodes;
+use bitcoin::hashes::_export::_core::fmt::Formatter;
+use bitcoin::util::taproot::TAPROOT_CONTROL_NODE_SIZE;
+use bitcoin::util::taproot::{
+    LeafVersion, TaprootBuilder, TaprootBuilderError, TaprootSpendInfo, TAPROOT_CONTROL_BASE_SIZE,
+};
+use bitcoin::{secp256k1, Script};
+use errstr;
+use expression::{self, FromTree, Tree};
+use miniscript::{limits::TAPROOT_MAX_NODE_COUNT, Miniscript};
+use std::cmp::{self, max};
+use std::sync::Arc;
+use std::{fmt, str::FromStr};
+use Tap;
+use {Error, MiniscriptKey};
+
+/// A Taproot Tree representation.
+// Hidden leaves are not yet supported in descriptor spec. Conceptually, it should
+// be simple to integrate those here, but it is best to wait on core for the exact syntax.
+#[derive(Clone, Ord, PartialOrd, Eq, PartialEq, Hash)]
+pub enum TapTree<Pk: MiniscriptKey> {
+    /// A taproot tree structure
+    Tree(Arc<TapTree<Pk>>, Arc<TapTree<Pk>>),
+    /// A taproot leaf denoting a spending condition
+    // A new leaf version would require a new Context, therefore there is no point
+    // in adding a LeafVersion with Leaf type here. All Miniscripts right now
+    // are of Leafversion::default
+    Leaf(Arc<Miniscript<Pk, Tap>>),
```

Comment:
I think the problem may be solved by having a dedicated enum variant for each Leaf version. So this will become `LeafTapScript`, the enum will become `#[non_exhaustive]` and there will be no problem to add more contexts with more enum variants with post-`0xC0` leafs

Reply:
Yeah, marking everything non-exhaustive is a big step. I suppose we will have a big breaking as soon as the MSRV lands